test/functional/operation_example.test.js

```diff
@@ -3599,68 +3599,6 @@ describe('Operation Examples', function() {
     }
   });
 
-  /**
-   * An example that shows how to force close a db connection so it cannot be reused.
-   *
-   * @example-class Db
-   * @example-method close
-   * @ignore
-   */
-  it('shouldCorrectlyFailOnRetryDueToAppCloseOfDb', {
-    metadata: {
-      requires: { topology: ['single', 'replicaset', 'sharded', 'ssl', 'heap', 'wiredtiger'] }
-    },
-
-    // The actual test we wish to run
-    test: function(done) {
-      const configuration = this.configuration;
-      if (configuration.usingUnifiedTopology()) {
-        // The new topology type has loose concepts of 'closing' and 'opening' a client. It will
-        // simply attempt here to retry the connection and reconnect, so this is a bad test for
-        // the driver in that configuration.
-
-        return this.skip();
-      }
-
-      var client = configuration.newClient(configuration.writeConcernMax(), { poolSize: 1 });
-      client.connect(function(err, client) {
-        // LINE var MongoClient = require('mongodb').MongoClient,
-        // LINE   test = require('assert');
-        // LINE const client = new MongoClient('mongodb://localhost:27017/test');
-        // LINE client.connect(function(err, client) {
-        // LINE   var db = client.db('test);
-        // REPLACE configuration.writeConcernMax() WITH {w:1}
-        // REMOVE-LINE restartAndDone
-        // REMOVE-LINE done();
-        // REMOVE-LINE var db = client.db(configuration.db);
-        // BEGIN
-        var db = client.db(configuration.db);
-        // Fetch a collection
-        var collection = db.collection('shouldCorrectlyFailOnRetryDueToAppCloseOfDb');
-
-        // Insert a document
-        collection.insertOne({ a: 1 }, configuration.writeConcernMax(), function(err, result) {
-          test.ok(result);
-          test.equal(null, err);
-
-          // Force close the connection
-          client.close(true, function(err) {
-            test.equal(null, err);
-
-            // Attemp to insert should fail now with correct message 'db closed by application'
-            collection.insertOne({ a: 2 }, configuration.writeConcernMax(), function(err, result) {
-              expect(err).to.exist;
-              expect(result).to.not.exist;
-
```

Comment:
This test attempts an operation against a closed client. This does fail with an error, as expected, but it also creates a session that never gets closed. We shouldn't be querying against a client that is closed, so we can delete the test.

Reply:
Now that I'm looking at this, I think we might want to actually keep this test. It's correctly returning an error because the client is closed, but why isn't the second call to `close` (L3655) clearing the session cache? My guess is because the implementation bails if the `topology == null`:
```
      if (client.topology == null) {
        return cb();
      }
```

I think what we actually need to do here is modify `MongoClient#close` to clear the session pool (if it exists) if the `topology == null`

Reply:
This is something I was actually confused about. There's a set of sessions stored on the `MongoClient` and a list of sessions stored in the session pool of the `topology`. The second `insertOne` op in this test creates a session in the (now-closed) topology's session pool. To make the test pass, it's not enough to attempt to change `MongoClient#close` to clear its own set of sessions (if it exists)-- we would need a reference to the topology to re-clear its session pool.